### PR TITLE
Add chapter about using secrets during a build

### PR DESCRIPTION
**Pull request by php-coder**
Documentation for https://github.com/openshift/origin/pull/6636

@mfojtik @bparees PTAL.

### CODE REVIEW

File: dev_guide/builds.adoc [hunk 1]
```diff
@@ -978,6 +978,139 @@ containing an inline Dockerfile that potentially replaces one within the
...
+in the final application image produced by the build.
+
+For example, you are building a NodeJS application and you set up your private
+mirror for NodeJS modules. In order to download modules from that private mirror,
+you have to supply a custom *_.npmrc_* file for the build that contains a URL,
+username and password. For security reasons, you do not want to expose your
+credentials in the application image.
+
+While this example describes NodeJS, you can use the same approach for adding
+SSL certificates into *_/etc/ssl/certs_* directory, API keys or tokens, license
```
Comment (bparees) [inline dev_guide/builds.adoc]: "into the /etc/ssl/certs directory."

File: dev_guide/builds.adoc [hunk 2]
```diff
@@ -978,6 +978,139 @@ containing an inline Dockerfile that potentially replaces one within the
...
+
+First, you have to create the `Secret` you want to use. You can do that by
+executing the following command:
+
+----
+$ oc secrets new secret-npmrc .npmrc=~/.npmrc
+----
+
+This command will create a new `Secret` named *_secret-npmrc_* and store
+the base64 encoded content of the *_~/.npmrc_* in it.
```
Comment (bparees) [inline dev_guide/builds.adoc]: "of the ~/.npmrc file in it."

File: dev_guide/builds.adoc [hunk 3]
```diff
@@ -978,6 +978,139 @@ containing an inline Dockerfile that potentially replaces one within the
...
+You can also specify the destination directory when creating a new `BuildConfig`:
+
+----
+$ oc new-build openshift/nodejs-010-centos7~https://github.com/openshift/nodejs-ex.git --build-secret “secret-npmrc:/etc”
+----
+
+In both cases, the *_.npmrc_* file will be added into the *_/etc_* directory of
+the build environment. Note that for a
+link:../architecture/core_concepts/builds_and_image_streams.html#docker-build[Docker strategy]
+the destination directory must be relative.
```
Comment (bparees) [inline dev_guide/builds.adoc]: "must be a relative path"

**Comment by bparees** [comment]
couple nits, otherwise lgtm.
ptal @openshift/team-documentation

File: dev_guide/builds.adoc [hunk 4]
```diff
@@ -978,6 +978,139 @@ containing an inline Dockerfile that potentially replaces one within the
...
+----
+
+If you want to create a new `BuildConfig` and you want to include the secrets
+in it, you can run the following command:
+
+----
+$ oc new-build openshift/nodejs-010-centos7~https://github.com/openshift/nodejs-ex.git --build-secret secret-npmrc
+----
+
+During the build in in both of these examples the *_.npmrc_* will be copied into
```
Comment (adellape) [inline dev_guide/builds.adoc]: "in in"

File: dev_guide/builds.adoc [hunk 5]
```diff
@@ -978,6 +978,139 @@ containing an inline Dockerfile that potentially replaces one within the
 contents of the binary archive.
 ====
 
+[[using-secrets]]
+== Using Secrets during a Build
+
+[[using-secrets-overview]]
+=== Overview
```
Comment (adellape) [inline dev_guide/builds.adoc]: I think the above two lines can just be removed.

File: dev_guide/builds.adoc [hunk 6]
```diff
@@ -978,6 +978,139 @@ containing an inline Dockerfile that potentially replaces one within the
...
+
+When you are using a `Source` strategy all defined source secrets will be copied
+to the respective `destinationDir`. If you left `destinationDir` empty then the
+secrets will be placed to the working directory of the builder image. The same
+rule is used when a destination directory is a relative path - the secrets will
+be placed into the paths that are relative to the image's working directory.
+The `destinationDir` must exist or an error will occur. No directory paths are
+created during the copy process.
+
+Keep in mind that in current implementation files with the secrets are
```
Comment (adellape) [inline dev_guide/builds.adoc]: "in the current implementation, files"

File: dev_guide/builds.adoc [hunk 7]
```diff
@@ -978,6 +978,139 @@ containing an inline Dockerfile that potentially replaces one within the
 contents of the binary archive.
 ====
 
+[[using-secrets]]
+== Using Secrets during a Build
```
Comment (adellape) [inline dev_guide/builds.adoc]: s/during/During/

File: dev_guide/builds.adoc [hunk 8]
```diff
@@ -978,6 +978,139 @@ containing an inline Dockerfile that potentially replaces one within the
...
+the layer where they were added.
+
+[[using-secrets-custom-strategy]]
+=== Custom Strategy
+
+When you are using a `Custom` strategy, then all the defined source secrets will
+be available inside the builder container under
+*_/var/run/secrets/openshift.io/build_* directory. It is the custom build image's
+responsibility to use these secrets appropriately. The `Custom` strategy also
+allows secrets to be defined as described link:#custom-secrets[here]. There is
```
Comment (adellape) [inline dev_guide/builds.adoc]:
```
as described in link:#custom-secrets[Secrets].
```

**Comment by adellape** [comment]
@php-coder Thanks. Some comments inline.

**Comment by php-coder** [comment]
@bparees @adellape Thanks for the feedback. I've fixed them.

**Comment by mfojtik** [comment]
LGTM

**Comment by mfojtik** [comment]
@adellape note that this feature is not merged yet, it is waiting for: https://github.com/openshift/origin/pull/6636

**Comment by adellape** [comment]
@php-coder Changes LGTM.

Waiting on code to merge.

**Comment by mfojtik** [comment]
@adellape this is now merged, feel free to merge this one.

**Comment by adellape** [comment]
[rev_history]
|link:../dev_guide/builds.html[Builds]
|Added a link:../dev_guide/builds.html#using-secrets[Using Secrets During a Build] section.
%

**Pull request merged**